Greatly optimise selective masking vector exports

nyalldawson opened the pull request:
Optimise the logic used when the new geometry backend for selective masking is in effect:

Whenever its SAFE, instead of calculating an "entire map" clipping path and then applying this for every feature being rendered,
we now defer the calculation of the clipping path until we are rendering individual features. Then, we create a clipping path
which contains ONLY the mask paths which are within the area being drawn over.

This avoids having the entire map clipping path being used for EVERY feature being rendered, which results in huge PDF/SVG
exports when masks are in effect, and instead results in clipping paths which are confined just to a sensible area
around each rendered feature.

In some complex test projects this reduces the PDF export size by a factor of 0.01, without any loss of quality or regressions in export time!! (Eg 60Mb -> 700kb). It also results in PDFs/SVGs which open much quicker in viewers and editors, and don't grind their operation to a halt.

Fixes https://github.com/qgis/QGIS/issues/50734
Fixes https://github.com/qgis/QGIS/issues/54788

nyalldawson committed:
Add method to collect symbol layer clip geometries intersecting given bounds

nyalldawson committed:
Add argument to QgsSymbol::startRender for whether symbol is subsymbol

Allows us to know in startRender whether the symbol is actually
a subsymbol for a QgsSymbolLayer. Currently unused.

nyalldawson committed:
Minor cleanup

nyalldawson committed:
Make installMasks return TRUE if any masks were found and installed

nyalldawson committed:
Add CanCalculateMaskGeometryPerFeature flag to "simple" symbol layers

These are the symbol layer classes where there's no special logic
required relating to feature rendering and features are rendered
one-by-one, with no sub symbols.

nyalldawson committed:
Greatly optimise selective masking vector exports

Optimise the logic used when the new geometry backend for
selective masking is in effect:

Whenever its SAFE, instead of calculating an "entire map" clipping
path and then applying this for every feature being rendered,
we now defer the calculation of the clipping path until we
are rendering individual features. Then, we create a clipping path
which contains ONLY the mask paths which are within the area
being drawn over.

This avoids having the entire map clipping path being used for
EVERY feature being rendered, which results in huge PDF/SVG
exports when masks are in effect, and instead results in
clipping paths which are confined just to a sensible area
around each rendered feature.

In some complex test projects this reduces the PDF export
size by a factor of 0.01!! (and results in PDFs/SVGs which
open much quicker in viewers and editors, and don't grind
their operation to a halt).

nyalldawson committed:
Update test masks

nyalldawson commented:
@troopa81 here's the next optimisation for selective masking exports. I consider this one the "holy grail", which squashes the last of the remaining big limitations of the vectorised selective masking.

troopa81 commented on src/core/symbology/qgssymbollayer.cpp:
```diff
@@ -1000,20 +1005,55 @@ void QgsSymbolLayer::prepareMasks( const QgsSymbolRenderContext &context )
+  }
+  else if ( rect.isValid() )
+  {
+    // find just the clip geometries within the area the symbol layer will be drawn over
```
It looks like pretty much the same piece of code of prepareMasks, can we factor it a bit in a method ?

nyalldawson replied:
Ok, done

troopa81 reviewed (commented):
I made a first quick review, I'll try to finish it in a few hours.

Having a tailored mask for each feature is interesting, but I think it brings a lot of complexity to the code (specially in QgsSymbol::renderFeature). 

Like I said before, I'm wondering if we could not have improved Qt by sharing/referening clippingPath when writing PDF. This way, we would have had one global mask (maybe built and simplified using the geometry paint device) applied to all features. In the end, the crash appears when serializing, not painting.

Therefore, the more feature you have, the bigger the file will be because you multiply masks, while with a global mask your will size stay constant.

nyalldawson commented on src/core/symbology/qgssymbollayerutils.h:
```diff
@@ -929,6 +929,18 @@ class CORE_EXPORT QgsSymbolLayerUtils
+     *
+     * \since QGIS 3.38
+     */
+    static QVector< QgsGeometry > collectSymbolLayerClipGeometries( const QgsRenderContext &context, const QString &symbolLayerId, const QRectF &bounds );
```
QgsRenderContext already is quite a large API -- I'd rather keep utility functions out of there as much as possible, and leave it focused on just getter/setter properties.

nyalldawson commented on src/core/symbology/qgssymbollayerutils.cpp:
```diff
@@ -5418,6 +5418,26 @@ void QgsSymbolLayerUtils::resetSymbolLayerIds( QgsSymbolLayer *symbolLayer )
+  clipGeometries.erase(
+    std::remove_if( clipGeometries.begin(), clipGeometries.end(), [&boundsRect]( const QgsGeometry & geometry )
+  {
+    return !geometry.boundingBoxIntersects( boundsRect );
```
@troopa81 

> Did you consider returning an intersection so we lower even more the size of the mask? or is it overkill and the intersection would be too much time consuming?

When geometries are added by the geometry paint device they are all split into singlepart geometry objects. That means that each geometry here will correspond to a single character (or symbol component), so I don't think we'll gain much advantage from doing an actual intersection here vs the cost of the intersection calculation itself.

nyalldawson committed:
Add optional flag to force global ("entire map") clipping paths

This is a new opt-in flag for map settings/render context/layouts.
If set, then when applying clipping paths for selective masking,
we always use global ("entire map") paths, instead of calculating
local clipping paths per rendered feature. This results in
considerably more complex vector exports in all current Qt versions,
but gives us a way to force this IF/when a future Qt version adds
optimisations which make global masks desirable.

nyalldawson commented:
@troopa81 
> I made a first quick review, I'll try to finish it in a few hours.

Thanks!

> Like I said before, I'm wondering if we could not have improved Qt by sharing/referening clippingPath when writing PDF. This way, we would have had one global mask (maybe built and simplified using the geometry paint device) applied to all features. 

> Therefore, the more feature you have, the bigger the file will be because you multiply masks, while with a global mask your will size stay constant.

That sounds good, and it would definitely be nice to see this happen in some future Qt release! :+1:  I've now added a render context/layout flag which forces the "entire map" clipping paths to always be used. It's opt in, but if Qt does improve the situation then we'll easily be able to switch back by just enabling this flag. (At some stage I'd love to make some advanced dialog for configuring layout export presets, and then we could even expose this setting to users who want fine control over this, and all the other mask simplification parameters too).

Pragmatically though, there's no related Qt improvements in current Qt 6 versions and use of local clip masks fixes the bulk of the issues with exporting complex layouts, so we should make this default behaviour for now.

> In the end, the crash appears when serializing, not painting.

It's caused during serializing the painter **state**, because the clip path is too complex for the structures Qt internally is using to do this. So we'd actually need two upstream improvements -- one would be the "write clip path once when exporting to SVG/PDF" improvement, and the other would be "adapt QPainter state storage to handle complex clip paths". 

On the other hand, with the changes here I can confirm that use of local clip paths fixes all the crashing projects from the linked tickets with current Qt stable versions.

nyalldawson committed:
Use SymbolRenderHint for flagging subsymbols

Instead of the extra argument to startRender

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -329,9 +329,10 @@ class CORE_EXPORT QgsSymbol
      * \see stopRender()
      */
-    void startRender( QgsRenderContext &context, const QgsFields &fields = QgsFields() );
+    void startRender( QgsRenderContext &context, const QgsFields &fields = QgsFields(), bool isSubSymbol = false );
```
@troopa81 indeed -- and it turns out there's a natural place for this already, in the symbols Qgis::SymbolRenderHint flags!

I've implemented this now.

nyalldawson committed:
Reduce duplicate code

troopa81 commented:
Is it normal ?
![layout_export_mask](https://github.com/qgis/QGIS/assets/14358135/2ceb554e-939f-485c-9e5a-a068bd3b71c4)

nyalldawson commented:
@troopa81 

> Is it normal

Yes, it's relating to that qt issue regarding paths bleeding out of the map frame's clip area. I'm fairly certain I can fix that now by doing the intersection of map item clip path and selective masking clip path via the geometry approach, but I'll do that in a different PR. And if it works I'll regenerate all the reference images so that none of them have the bleed anymore.

troopa81 commented on src/core/symbology/qgsarrowsymbollayer.cpp:
```diff
@@ -215,7 +215,9 @@ void QgsArrowSymbolLayer::startRender( QgsSymbolRenderContext &context )
   mComputedArrowType = arrowType();
 
-  mSymbol->startRender( context.renderContext() );
+  mSymbol->setRenderHints( mSymbol->renderHints() | Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
Wouldn't be better to do it when adding the symbol to its parent symbol layer so the flag could be reused outside of the rendering pipeline?

nyalldawson replied:
Let's defer that till it's actually needed. Turns out it's not so trivial to do, as the render hints aren't cloned (and I don't think they should be, either). The logic ends up much more complex as a result.

troopa81 commented on src/core/symbology/qgssymbol.cpp:
```diff
@@ -1642,6 +1663,23 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
+    // if this symbol layer has associated clip masks, we need to render it to a QPicture first so that we can
+    // determine the actual rendered bounds of the symbol. We'll then use that to retrieve the clip masks we need
+    // to apply when painting the symbol via this QPicture.
+    const bool hasClipGeometries = !context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
```
Shall we not test `!mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol` ? Isn't that the opposite of the condition above?

nyalldawson replied:
Right, fixed

nyalldawson committed:
Fix check for sub symbol

nyalldawson commented on src/core/symbology/qgssymbol.cpp:
```diff
@@ -1633,6 +1654,9 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
     mSymbolRenderContext->expressionContextScope()->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_layer_count" ), mLayers.count(), true ) );
 
+  const bool maskGeometriesDisabledForSymbol = context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
+                                               && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
```suggestion
    && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```

nyalldawson committed:
Update src/core/symbology/qgssymbol.cpp

nyalldawson commented on src/core/symbology/qgssymbol.cpp:
```diff
@@ -1633,6 +1654,9 @@ void QgsSymbol::renderFeature( const QgsFeature &feature, QgsRenderContext &cont
     mSymbolRenderContext->expressionContextScope()->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_layer_count" ), mLayers.count(), true ) );
 
+  const bool maskGeometriesDisabledForSymbol = context.testFlag( Qgis::RenderContextFlag::AlwaysUseGlobalMasks )
+    && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```
```suggestion
      && !mRenderHints.testFlag( Qgis::SymbolRenderHint::IsSymbolLayerSubSymbol );
```

nyalldawson committed:
Update src/core/symbology/qgssymbol.cpp
